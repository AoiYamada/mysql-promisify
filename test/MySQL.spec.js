'use strict';

const mysql = require('mysql');
const { MySQL } = require('../src/MySQL');

describe('MySQL', () => {
  const db = new MySQL({
    host: '127.0.0.1',
    port: 5001,
    user: 'root',
    password: 'rootpassword',
    charset: 'utf8',
    database: 'test',
    timeout: 60000,
    // multipleStatements: true,
  });

  beforeAll(async () => {
    await db.query({
      sql: `
        DROP TABLE IF EXISTS test;
      `,
    });

    await db.query({
      sql: `
        CREATE TABLE test (
          id INT(11) NOT NULL AUTO_INCREMENT PRIMARY KEY,
          testcol1 VARCHAR(45) NULL DEFAULT NULL,
          testcol2 VARCHAR(45) NULL DEFAULT NULL,
          testcol3 VARCHAR(45) NULL DEFAULT NULL,
          testcol4 VARCHAR(45) NULL DEFAULT NULL,
          testcol5 VARCHAR(45) NULL DEFAULT NULL
        );
      `,
    });
  });

  afterAll(async () => {
    await db.query({ sql: 'TRUNCATE `test`.`test`;' });
    await db.end();
  });

  test('Create Data to DB', async () => {
    const result = await db.query({
      sql: `
        INSERT INTO
          test
          (
            testcol1,
            testcol2,
            testcol3
          )
        VALUES
          (
            :val1,
            :val2,
            :val3
          )
        ;
      `,
      params: {
        val1: 123,
        val2: 456,
        val3: 789,
      },
    });

    expect(result).toMatchSnapshot({
      results: {
        insertId: expect.any(Number),
      },
    });
  });

  test('Read Data From DB', async () => {
    const result = await db.query({
      sql: 'SELECT * FROM test;',
    });

    expect(result).toMatchSnapshot();
  });

  test('Use escape string as column name or table name', async () => {
    const table = 'test';
    const result = await db.query({
      sql: 'SELECT * FROM |table;',
      params: {
        table,
      },
    });

    expect(result).toMatchSnapshot();
  });

  test('Get the substituded sql', () => {
    const data = {
      sql: `
        INSERT INTO
          test
          (
            testcol1,
            testcol2,
            testcol3
          )
        VALUES
          (
            :val1,
            :val2,
            :val3
          )
        ;
    `,
      params: {
        val1: 123,
        val2: 456,
        val3: 789,
      },
    };

    const query = MySQL.queryFormat(data.sql, data.params);

    expect(query).toBe(`
        INSERT INTO
          test
          (
            testcol1,
            testcol2,
            testcol3
          )
        VALUES
          (
            ${mysql.escape(data.params.val1)},
            ${mysql.escape(data.params.val2)},
            ${mysql.escape(data.params.val3)}
          )
        ;
    `);
  });

  describe('Transaction', () => {
    test('Rollback', async () => {
      const tdb = await db.getTransactionDb();

      expect(tdb).toMatchSnapshot({
        _connection: expect.any(Object),
      });

      const { results: readResultsBefore } = await db.query({
        sql: 'SELECT * FROM test;',
      });

      const { results: insertResults } = await tdb.query({
        sql: `
          INSERT INTO
            test
            (
              testcol1,
              testcol2,
              testcol3
            )
          VALUES
            (
              :val1,
              :val2,
              :val3
            )
          ;
        `,
        params: {
          val1: 123,
          val2: 456,
          val3: 789,
        },
      });

      expect(insertResults).toMatchSnapshot();
      await tdb.rollback();
      expect(tdb._connection).toBe();

      const { results: readResultsAfter } = await db.query({
        sql: 'SELECT * FROM test;',
      });

      expect(readResultsBefore.length).toBe(readResultsAfter.length);
    });
    test('Commit', async () => {
      const tdb = await db.getTransactionDb();

      expect(tdb).toMatchSnapshot({
        _connection: expect.any(Object),
      });

      const { results: readResultsBefore } = await db.query({
        sql: 'SELECT * FROM test;',
      });

      const { results: insertResults } = await tdb.query({
        sql: `
          INSERT INTO
            test
            (
              testcol1,
              testcol2,
              testcol3
            )
          VALUES
            (
              :val1,
              :val2,
              :val3
            )
          ;
        `,
        params: {
          val1: 123,
          val2: 456,
          val3: 789,
        },
      });

      expect(insertResults).toMatchSnapshot();
      await tdb.commit();
      expect(tdb._connection).toBe();

      const { results: readResultsAfter } = await db.query({
        sql: 'SELECT * FROM test;',
      });

      expect(readResultsAfter.length).toBe(readResultsBefore.length + 1);
    });
  });
});
