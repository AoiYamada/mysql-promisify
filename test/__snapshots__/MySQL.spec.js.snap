// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySQL Create Data to DB 1`] = `
Object {
  "fields": undefined,
  "results": Object {
    "affectedRows": 1,
    "changedRows": 0,
    "fieldCount": 0,
    "insertId": Any<Number>,
    "message": "",
    "protocol41": true,
    "serverStatus": 2,
    "warningCount": 0,
  },
}
`;

exports[`MySQL Read Data From DB 1`] = `
Object {
  "fields": Array [
    FieldPacket {
      "catalog": "def",
      "charsetNr": 63,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 16899,
      "length": 11,
      "name": "id",
      "orgName": "id",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 3,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol1",
      "orgName": "testcol1",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol2",
      "orgName": "testcol2",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol3",
      "orgName": "testcol3",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol4",
      "orgName": "testcol4",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol5",
      "orgName": "testcol5",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
  ],
  "results": Array [
    RowDataPacket {
      "id": 1,
      "testcol1": "123",
      "testcol2": "456",
      "testcol3": "789",
      "testcol4": null,
      "testcol5": null,
    },
  ],
}
`;

exports[`MySQL Transaction Commit 1`] = `
Object {
  "_connection": Any<Object>,
}
`;

exports[`MySQL Transaction Commit 2`] = `
OkPacket {
  "affectedRows": 1,
  "changedRows": 0,
  "fieldCount": 0,
  "insertId": 3,
  "message": "",
  "protocol41": true,
  "serverStatus": 3,
  "warningCount": 0,
}
`;

exports[`MySQL Transaction Rollback 1`] = `
Object {
  "_connection": Any<Object>,
}
`;

exports[`MySQL Transaction Rollback 2`] = `
OkPacket {
  "affectedRows": 1,
  "changedRows": 0,
  "fieldCount": 0,
  "insertId": 2,
  "message": "",
  "protocol41": true,
  "serverStatus": 3,
  "warningCount": 0,
}
`;

exports[`MySQL Use escape string as column name or table name 1`] = `
Object {
  "fields": Array [
    FieldPacket {
      "catalog": "def",
      "charsetNr": 63,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 16899,
      "length": 11,
      "name": "id",
      "orgName": "id",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 3,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol1",
      "orgName": "testcol1",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol2",
      "orgName": "testcol2",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol3",
      "orgName": "testcol3",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol4",
      "orgName": "testcol4",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
    FieldPacket {
      "catalog": "def",
      "charsetNr": 33,
      "db": "test",
      "decimals": 0,
      "default": undefined,
      "flags": 0,
      "length": 135,
      "name": "testcol5",
      "orgName": "testcol5",
      "orgTable": "test",
      "protocol41": true,
      "table": "test",
      "type": 253,
      "zeroFill": false,
    },
  ],
  "results": Array [
    RowDataPacket {
      "id": 1,
      "testcol1": "123",
      "testcol2": "456",
      "testcol3": "789",
      "testcol4": null,
      "testcol5": null,
    },
  ],
}
`;
